anomalib/utils/callbacks/visualizer_base.py

```diff
@@ -0,0 +1,106 @@
+        show_images: bool = False,
+        log_images: bool = True,
+        save_images: bool = True,
+    ):
+        """Visualizer callback."""
+        if mode not in ["full", "simple"]:
+            raise ValueError(f"Unknown visualization mode: {mode}. Please choose one of ['full', 'simple']")
+        self.mode = mode
+        if task not in ["classification", "segmentation"]:
+            raise ValueError(f"Unknown task type: {mode}. Please choose one of ['classification', 'segmentation']")
+        self.task = task
+        self.inputs_are_normalized = inputs_are_normalized
+        self.show_images = show_images
+        self.log_images = log_images
+        self.save_images = save_images
+        self.image_save_path = Path(image_save_path)
+
+    def _add_to_logger(
+        self,
+        image: np.ndarray,
+        module: AnomalyModule,
+        trainer: pl.Trainer,
+        filename: Union[Path, str],
+    ):
+        """Log image from a visualizer to each of the available loggers in the project.
+
+        Args:
+            image (np.ndarray): Image that should be added to the loggers.
+            module (AnomalyModule): Anomaly module.
+            trainer (Trainer): Pytorch Lightning trainer which holds reference to `logger`
+            filename (Path): Path of the input image. This name is used as name for the generated image.
+        """
+        # Store names of logger and the logger in a dict
+        available_loggers = {
+            type(logger).__name__.lower().rstrip("logger").lstrip("anomalib"): logger for logger in trainer.loggers
+        }
+        # save image to respective logger
+        if self.log_images:
+            for log_to in available_loggers:
+                # check if logger object is same as the requested object
+                if isinstance(available_loggers[log_to], ImageLoggerBase):
+                    logger: ImageLoggerBase = cast(ImageLoggerBase, available_loggers[log_to])  # placate mypy
+                    if isinstance(filename, Path):
+                        _name = filename.parent.name + "_" + filename.name
+                    elif isinstance(filename, str):
+                        _name = filename
+                    logger.add_image(
+                        image=image,
+                        name=_name,
+                        global_step=module.global_step,
+                    )
+
+    def on_test_end(self, trainer: pl.Trainer, pl_module: AnomalyModule) -> None:
+        """Sync logs.
+
+        Currently only ``AnomalibWandbLogger.save`` is called from this method.
+        This is because logging as a single batch ensures that all images appear as part of the same step.
+
+        Args:
+            trainer (pl.Trainer): Pytorch Lightning trainer
+            pl_module (AnomalyModule): Anomaly module (unused)
+        """
+        for logger in trainer.loggers:
+            if isinstance(logger, AnomalibWandbLogger):
+                logger.save()
```

Comment:
Unsure whether it's "bad practice" to call .save on a `AnomalibWandbLogger` multiple times, as is done currently. Any insights?

Reply:
It has been a while since I have looked into AnomalibWandbLogger but ideally it should only be called once at the very end. This is explained in the comment above `This is because logging as a single batch ensures that all images appear as part of the same step.`. In case it is being called multiple times then my guess is that the `isinstance` fallsback to the base class and probably logger ends up always being an instance of `AnomalibWandbLogger` (`LightningLogger` more specifically). Maybe a good idea would be to use `type()` but mypy or pylint might complain.

Reply:
I opted with clearing `self.images` on save. Thus, all predictions are in the first step, and all metrics in the second step